### DISCUSSION TITLE
Помощь с роутерами

### DISCUSSION THREAD

**Discussion by Waujito** [Waujito/youtubeUnblock#172] [General]
Вопросы по роутерам, архитектурам, запуску на роутерах и так далее можете задавать здесь, чтобы не плодить лишних issues аля "подскажите архитектуру для mips"

**Comment by MayanKoyote** [comment]
Доброго времени.
Собираю версию ramips/rt305x (mipsel_24kc) для Openwrt18.06.9 через SDK по инструкции из README.
Есть подозрение что в результате я получаю состояние на 58f4802 .
Вопрос: если я хочу получить более актуальную версию (562f17e) нужно ли мне в файле [youtubeUnblock/Makefile](https://github.com/Waujito/youtubeUnblock/blob/e764a20baa69eeba5b36d0a847435bcf9c6efaf2/youtubeUnblock/Makefile#L7) менять `58f4802f6466e48734d303ff0a2deb2bfe9e7f16` на `562f17e932547f9d92111d6c9245c6170931ca1d` ?

**Reply by Waujito** [reply]
Да, я мог забыть поменять хэши. Вы можете склонить ветку openwrt куда-нибудь рядом с SDK и поменять хэш в мэйкфайле на тот, который нужен. В фидах написать src-local youtubeUnblock /path/to/directory, как тут: https://openwrt.org/docs/guide-developer/helloworld/chapter4

**Comment by r0otx** [comment]
Приветствую! Большое спасибо за пакет и труды, в такое непростое время, как свет в конце тоннеля подобные продукты :) Но  я не могу понять запускается пакет или нет, если я запускаю youtubeUnblock, youtube не загружает видео. Я делаю следующие шаги:

1. Перехожу в ssh на /opt/etc/init.d/S91youtubeUnblock и открываю для редактирования.
2. Добавляю флаги в ARGS="--fake-sni-type=custom --fake-custom-payload=00 --fake-sni-seq-len=10 --frag-sni-faked=1 --frag-sni-reverse=0 --frag-sni-pos=1 --frag-middle-sni=0 --seg2delay=100"
3. После этого, запускаю youtubeUnblock &
4. В консоли вижу этот текст:

![image](https://github.com/user-attachments/assets/20e53577-30c4-4248-ae69-417e1ea89310)

5. YouTube на ТВ не работает :( Что я делаю не так?

Запускаю на entware, роутер ASUS GT-AX11000 на прошивке Мерлина.

# uname -a
Linux ax-router 4.1.51 #4 SMP PREEMPT Wed Jul 31 20:57:41 EDT 2024 aarch64 ASUSWRT-Merlin

Заранее огромное спасибо за помощь!

**Reply by Waujito** [reply]
Пока вы в консоли должен отображаться вывод о том, что анблок видит ютуб. Если этого вывода нет - что-то с правилами фаервола. Отправьте `iptables -t mangle -nvL` сюда. Также можно попробовать добавить правила фаервола самостоятельно (просто вводя строчки из README в терминале). Также убедитесь что в браузере отключены QUIC и kyber. Для ТВ поможет добавление флага (--quic-block)

**Reply by r0otx** [reply]
<img width="583" alt="image" src="https://github.com/user-attachments/assets/caf8dc56-b55b-4f20-bf3f-624b044ee0aa">
вот что выводит iptables после запуска пакета

**Reply by Waujito** [reply]
А ну понятно. Надо запускать не `youtubeUnblock &`, а `/opt/etc/init.d/youtubeUnblock start`

**Reply by r0otx** [reply]
По полному пути пишет

-sh: /opt/etc/init.d/youtubeUnblock: not found

да и в директории этой лежит только S91youtubeUnblock, если запускаю его с командой start, то та же ошибка -sh: S91youtubeUnblock: not found хотя я в той же директории нахожусь. По полному пути до S91youtubeUnblock так же не найдено. Как-то не могу до сих пор понять, как этот энтвар работает, в традиционном линухе все сильно проще, а тут симлинк в системе есть, а запускается все через одно месте, с v2ray плагином тоже намучился, так и не запустил, конфиг правлю, а изменения не применяются, как-будто все из другого места стартует =/

**Reply by Waujito** [reply]
> По полному пути пишет
> 
> -sh: /opt/etc/init.d/youtubeUnblock: not found
> 
> да и в директории этой лежит только S91youtubeUnblock, если запускаю его с командой start, то та же ошибка -sh: S91youtubeUnblock: not found хотя я в той же директории нахожусь. По полному пути до S91youtubeUnblock так же не найдено. Как-то не могу до сих пор понять, как этот энтвар работает, в традиционном линухе все сильно проще, а тут симлинк в системе есть, а запускается все через одно месте, с v2ray плагином тоже намучился, так и не запустил, конфиг правлю, а изменения не применяются, как-будто все из другого места стартует =/

Вы как будто неправильно скрипт запускаете надо написать `/opt/etc/init.d/S91youtubeUnblock start`. или что то в этом роде. В целом еще можно запустить руками без скрипта, но надо будет добавить правила iptables как в readme.

**Comment by whood** [comment]
Здравствуйте. Пытаюсь завести на ZBT WE826-T2 [MT7620 MIPS 24K], но всё не получается.
Собственно собрал прошивку из репозитория https://gitlab.com/hadzhioglu/padavan-ng с правками конфига для добавления модулей.
Пробовал запускать и через статик и через Entware.

Прописал строки iptables из ридми, вот что выдает при попытке запуска.
```
/opt/home/admin # /etc/storage/youtubeUnblock
youtubeUnblock
Bypasses deep packet inspection systems that relies on SNI

GSO is enabled
IPv6 is enabled
Detected 1 config sections
The sections will be processed in ordred they goes in this outputSection #0
Using TCP segmentation
Fake SNI will be sent before each target client hello
Fragmentation Client Hello will be reversed
Past seq faking strategy will be used
Target sni domains: googlevideo.com,ggpht.com,ytimg.com,youtube.com,play.google.com,youtu.be,googleapis.com,googleusercontent.com,gstatic.com,l.google.com
mnl_socket_open: Protocol not supported
Unable to open socket: Protocol not supported
Thread 0 exited with status -1
```

Если прописать:
```
/opt/home/admin # modprobe nfnetlink
/opt/home/admin # modprobe nfnetlink_queue
/opt/home/admin # modprobe xt_connbytes
/opt/home/admin # modprobe xt_NFQUEUE
```

То получаем такой результат:
```
/opt/home/admin # /etc/storage/youtubeUnblock
youtubeUnblock
Bypasses deep packet inspection systems that relies on SNI

GSO is enabled
IPv6 is enabled
Detected 1 config sections
The sections will be processed in ordred they goes in this outputSection #0
Using TCP segmentation
Fake SNI will be sent before each target client hello
Fragmentation Client Hello will be reversed
Past seq faking strategy will be used
Target sni domains: googlevideo.com,ggpht.com,ytimg.com,youtube.com,play.google.                     com,youtu.be,googleapis.com,googleusercontent.com,gstatic.com,l.google.com
Queue 537 started
```
Грузиться ничего лучше не начинает при этом.
Ещё вот результат команды iptables -t mangle -nvL и lsmod. Не понимаю почему тут ошибка.
![putty_aFOURZSqU1](https://github.com/user-attachments/assets/f3dac115-9a87-49dd-a12f-ee07b7b8bb1a)

Честно говоря я в этом деле ничего не понимаю. Полдня потратил на чтение форумов и гайдов и такой затык по итогу.
До этого пользовался Openwrt с предустановленым вашим скриптом, там всё работало идеально.
Очень бы хотелось разобраться т.к. Падаван на моём устройстве выдаёт скорости в полтора-два раза выше чем openwrt.

**Reply by whood** [reply]
Со второго захода разобрался. Нужно было использовать более новую версию отсюда https://github.com/Waujito/youtubeUnblock/actions/runs/11570571161. Теперь всё работает прекрасно.

**Comment by MayanKoyote** [comment]
Вопрос по `kmod_youtubeUnblock`:
я правильно понимаю, что `kmod_youtubeUnblock` для работы не нужны ни `kmod-nfnetlink-queue` (и прочие зависимости из [OpenWRT pre configuration](https://github.com/Waujito/youtubeUnblock?tab=readme-ov-file#openwrt-pre-configuration)) ни правила iptables/nftables? Или что-то всё таки нужно?

**Reply by Waujito** [reply]
Всё верно. Нужно только третье ядро или больше

**Comment by r0otx** [comment]
А если правила в ручную в iptables заполнять, то необходимость в запуске службы отпадает полностью или они работают в совокупности?

**Reply by Waujito** [reply]
Помимо правил должен висеть процесс самой программы. Это можно сделать сервисом, а можно и нохопом, даемонайзом или чем нибудь другим.

**Reply by r0otx** [reply]
Разобрался, я пытался запустить сервис зайдя в папку init.d и почему-то именно в ней он не видел его. А когда делал то же самое с указанием полного пути /opt/etc/init.d/S90youtubeUnblock stop/start то все заработало и правила сами в iptables встали и аргументы подцепились, при этом редактировать этот файл давало, мистика....

**Comment by cdies** [comment]
Всем привет! У меня есть Raspberry PI 3B+, я хочу её использовать как шлюз по-умолчанию (Default gateway), чтобы на ней запустить youtubeUnblock и уже запрос с подменённым dpi отправлять в роутер. Пробовал настраивать iptables для этого, но никак не получается. Подскажите пожайлуста, может кто сталкивался с такой схемой, если так можете подсказать решение или хотя-бы направить в нужную сторону?

**Reply by Waujito** [reply]
Если хочется средствами сети это сделать, то по идее адрес распберри как шлюза (`ip route add default via 192.168.0.100 dev wlan0`) должен работать (плюс FORWARD ACCEPT на самой малинке)

**Reply by cdies** [reply]
Да, спасибо, получилось разобраться!

**Reply by cdies** [reply]
Описал в отдельной статье, что получилось в итоге: https://habr.com/ru/articles/871460/
Вкратце, что нужно сделать: в файле `/etc/sysctl.conf ` раскомментировать `net.ipv4.ip_forward=1` и в файле [`youtubeUnblock.service`](https://github.com/Waujito/youtubeUnblock/blob/main/youtubeUnblock.service) добавить команд для iptables в `ExecStartPre` и `ExecStop`:
```
[Unit]
Description=youtubeUnblock

[Service]
StandardError=journal
StandardOutput=journal
StandardInput=null
ExecStartPre=/bin/bash -c 'iptables -t mangle -A POSTROUTING -p tcp --dport 443 -j NFQUEUE --queue-num 537 --queue-bypass && iptables -t mangle -A OUTPUT -p tcp -m tcp --dport 443 -m connbytes --connbytes-dir original --connbytes-mode packets --connbytes 0:19 -j NFQUEUE --queue-num 537 --queue-bypass'
ExecStart=$(PREFIX)/bin/youtubeUnblock
ExecStop=/bin/bash -c 'iptables -t mangle -D POSTROUTING -p tcp --dport 443 -j NFQUEUE --queue-num 537 --queue-bypass && iptables -t mangle -D OUTPUT -p tcp -m tcp --dport 443 -m connbytes --connbytes-dir original --connbytes-mode packets --connbytes 0:19 -j NFQUEUE --queue-num 537 --queue-bypass'

[Install]
WantedBy=multi-user.target
```
Оставлю здесь краткое описание, на всякий случай, может кому понадобится.

**Comment by cactuslime1** [comment]
Доброго времени суток, установил последнюю версию анблока на openwrt, возникла ситуация: на беспроводных клиентах все работает, на проводных нет, подскажите куда посмотреть, что поменять? Спасибо

**Reply by Waujito** [reply]
Возможно, проблема в offloading. Посмотрите этот раздел ридми https://github.com/Waujito/youtubeUnblock?tab=readme-ov-file#nat-hardwaresoftware-offloading

**Reply by cactuslime1** [reply]
Действительно, проблема решилась. Спасибо

**Comment by MayanKoyote** [comment]
При сборке модуля ядра (v1.0.0-r10) в SDK OpenWrt (23.05.5, 24.10.0, оба ramips mt7620) нормально получать только ошибку:
```
Package kmod-youtubeUnblock is missing dependencies for the following libraries:
nf_conntrack.ko
```
или нужно сразу собирать с отключенным conntrack, используя `make kmake EXTRA_CFLAGS="-DNO_CONNTRACK"` ?
Или всё же не нормально?
Не то что бы он (conntrack) мне был нужен, но он же не отключен сразу (как я понимаю), а только упомянут как опциональный.
Кстати, как (и где?) должна полностью выглядеть эта строка с kmake?

Обновление 1: добавил в `youtubeUnblock/kyoutubeUnblock/Makefile` строку `DEPENDS:=+kmod-nf-conntrack`, модуль собрался без видимых при сборке ошибок. Работоспособность полученного модуля пока не тестировал.

Обновление 2: если вместо добавления строки `DEPENDS:=+kmod-nf-conntrack` в `youtubeUnblock/kyoutubeUnblock/Makefile` заменить `FILES:=$(PKG_BUILD_DIR)/kyoutubeUnblock.$(LINUX_KMOD_SUFFIX)` на
```
	FILES:= \
		$(PKG_BUILD_DIR)/kyoutubeUnblock.$(LINUX_KMOD_SUFFIX) \
		$(LINUX_DIR)/net/netfilter/nf_conntrack.ko \
		$(LINUX_DIR)/net/ipv4/netfilter/nf_defrag_ipv4.ko \
		$(LINUX_DIR)/net/ipv6/netfilter/nf_defrag_ipv6.ko
```
то модуль собирается тоже без видимых при сборке ошибок. Сам получаемый при этом файл `kyoutubeUnblock.ko` абослютно идентичен файлу `kyoutubeUnblock.ko` полученному через `DEPENDS:=+kmod-nf-conntrack` (контрольные суммы совпадают), только в ipk-пакет добавляются еще и файлы `nf_conntrack.ko`, `nf_defrag_ipv4.ko` и `nf_defrag_ipv6.ko`. Работоспособность полученного модуля пока не тестировал.

Обновление 3: попробовал собрать модуль в SDK 18.06.9, ramips rt305x и ramips mt7620 (этот для спортивного интереса). Собирал как есть, никаких изменений не вносил. Модуль в обоих случаях собрался без всякой любой ругани. Работоспособность полученного модуля пока не тестировал.

Всё это занятно, но что то тут не так.

**Reply by Waujito** [reply]
Добрый день!
Спасибо за подробный ресёрч. 
Да, в модуле появилась поддержка коннтрека. Она хоть и опциональная, но желательная. Если он ругается, лучше тогда я добавлю контрек в зависимости. Всё равно на обычных роутерах он обязательно должен быть из-за NAT.

**Comment by diks13** [comment]
Здравствуйте
Роутер Keenetic Giga KN-1012
Установил Entware
Пытаюсь поставить youtubeUnblock-1.0.0-10-f37c3dd-aarch64_cortex-a53-openwrt-23.05.ipk
Читаю инструкцию и не могу понять, как поставить ? 
Мне это сложно дается, нуждаюсь в помощи.

**Reply by Waujito** [reply]
Для entware надо использовать пакет с пометкой entware, а не openwrt. Архитектуру смотрите в `cat /proc/cpuinfo`

**Reply by diks13** [reply]
> Для entware надо использовать пакет с пометкой entware, а не openwrt. Архитектуру смотрите в `cat /proc/cpuinfo`

Установить получилось такими командами.
В консоли ввожу:
`exec sh`
Получаю
` BusyBox v1.37.0 (2025-01-13 14:28:09 UTC) built-in shell (ash)`
Ввожу
`opkg install youtubeUnblock-1.0.0-10-f37c3dd-entware-aarch64-3.10.ipk`
Получаю
`Package youtubeUnblockEntware (1.0.0-10-f37c3dd~f37c3dd-r10) installed in root is up to date.`
Ввожу
`/opt/etc/init.d/S91youtubeUnblock start.`
Получаю
`Usage: /opt/etc/init.d/S91youtubeUnblock {start|stop|restart|status|firewall-load|firewall-stop|init-system}`
Установка как я понял прошла успешно, выскочило меню запуска и на этом я застрял.
Ютуб проверил, без изменений((
Нуждаюсь в ваших подсказках.

**Reply by bursoft-git** [reply]
Добрый день!
Спасибо за отличную программу, развернул её на openwrt, всё заработало на дефолтных настройках! Супер!
Попробовал развернуть её на другом роутере keenetic kn-1810: установил entware, iptables, kmod, собственно сам анблок версии 1.0.0.
Проверяю работоспособность:
```
BusyBox v1.37.0 (2025-06-01 14:50:09 UTC) built-in shell (ash)

~ # /opt/etc/init.d/S91youtubeUnblock status
stopped
~ # curl -o/dev/null -k --connect-to ::google.com -k -L -H Host:\ mirror.gcr.io
https://test.googlevideo.com/v2/cimg/android/blobs/sha256:6fd8bdac3da660bde7bd0b
6f2b6a46e1b686afb74b9a4614def32532b73f5eaa
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:10 --:--:--     0^C
~ # /opt/etc/init.d/S91youtubeUnblock start
 Started youtubeUnblock
~ # curl -o/dev/null -k --connect-to ::google.com -k -L -H Host:\ mirror.gcr.io
https://test.googlevideo.com/v2/cimg/android/blobs/sha256:6fd8bdac3da660bde7bd0b
6f2b6a46e1b686afb74b9a4614def32532b73f5eaa
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1144  100  1144    0     0    778      0  0:00:01  0:00:01 --:--:--  2465
 11  664M   11 78.4M    0     0  5064k      0  0:02:14  0:00:15  0:01:59 5524k^C
```

C выключенным скриптом скорости нет, со включенным - есть, то есть вроде работает.
Но это на роутере, а на компьютере вообще не работает, ни по проводу, ни по беспроводу, независимо от браузера и настроек QUIC/Kyber
Подскажите, что можно проверить/настроить ещё, моих знаний не хватает(
Спасибо!

**Reply by Waujito** [reply]
> C выключенным скриптом скорости нет, со включенным - есть, то есть вроде работает. Но это на роутере, а на компьютере вообще не работает, ни по проводу, ни по беспроводу, независимо от браузера и настроек QUIC/Kyber Подскажите, что можно проверить/настроить ещё, моих знаний не хватает( Спасибо!

Смотрите по логам анблока как и куда идёт трафик. 
Можете скинуть сюда вывод `iptables -t mangle -nvL`
Чтобы смотреть логи можно анблок запустить просто командой `/opt/bin/youtubeUnblock`, предварительно остановив инстанс из скрипта. Также можно добавить флаг `--trace` чтобы получить более развёрнутые логи.

**Reply by bursoft-git** [reply]
Вот вывод `iptables -t mangle -nvL`
[iptables -t mangle -nvL.txt](https://github.com/user-attachments/files/21182221/iptables.-t.mangle.-nvL.txt)

Остановил инстанс, запустил `/opt/bin/youtubeUnblock --trace`, попробовал пооткрывать ютуб, но вывода не дождался.
```
~ # /opt/bin/youtubeUnblock --trace
youtubeUnblock 1.0.0-10
Bypasses deep packet inspection systems that rely on SNI

Running with flags: --queue-num=537 --threads=1 --packet-mark=32768 --trace --tls=enabled --frag=tcp --frag-sni-reverse=1 --frag-sni-faked=0 --frag-middle-sni=1 --frag-sni-pos=1 --fk-winsize=0 --fake-sni=1 --fake-sni-seq-len=1 --fake-sni-type=default --faking-strategy=pastseq --fake-seq-offset=10000 --seg2delay=0 --sni-domains=<trie of 157 vertexes> --sni-detection=parse --synfake=0 --udp-filter-quic=disabled
Queue 537 started
^CyoutubeUnblock stats: catched 0 packets, processed 0 packets, targetted 0 packets, sent over socket 0 packets
```

Не уверен, что делаю всё правильно (точнее, уверен, что делаю что-то не так, понять бы что))

**Reply by Waujito** [reply]
> Вот вывод `iptables -t mangle -nvL`
> [iptables -t mangle -nvL.txt](https://github.com/user-attachments/files/21182221/iptables.-t.mangle.-nvL.txt)
> 
> Остановил инстанс, запустил `/opt/bin/youtubeUnblock --trace`, попробовал пооткрывать ютуб, но вывода не дождался.
> ```
> ~ # /opt/bin/youtubeUnblock --trace
> youtubeUnblock 1.0.0-10
> Bypasses deep packet inspection systems that rely on SNI
> 
> Running with flags: --queue-num=537 --threads=1 --packet-mark=32768 --trace --tls=enabled --frag=tcp --frag-sni-reverse=1 --frag-sni-faked=0 --frag-middle-sni=1 --frag-sni-pos=1 --fk-winsize=0 --fake-sni=1 --fake-sni-seq-len=1 --fake-sni-type=default --faking-strategy=pastseq --fake-seq-offset=10000 --seg2delay=0 --sni-domains=<trie of 157 vertexes> --sni-detection=parse --synfake=0 --udp-filter-quic=disabled
> Queue 537 started
> ^CyoutubeUnblock stats: catched 0 packets, processed 0 packets, targetted 0 packets, sent over socket 0 packets
> ```
> 
> Не уверен, что делаю всё правильно (точнее, уверен, что делаю что-то не так, понять бы что))

Анблок в статистике написал что ничего не поймал. Возможно, после остановки инстанса правила удалились, и вывод команды с iptables изменился. Перед запуском сделайте `/opt/etc/init.d/S91youtubeUnblock firewall-load`, и убедитесь, что цепочка YOUTUBEUNBLOCK загружена, и что счётчики растут с каждым пакетом. Тогда можно запускать анблок и смотреть на трейс логи.

**Reply by bursoft-git** [reply]
Сделал `/opt/etc/init.d/S91youtubeUnblock firewall-load`, потом `/opt/bin/youtubeUnblock --trace`, попробовал открыть ютуб.
Получился вот такой вывод
[youtubeUnblock --trace.txt](https://github.com/user-attachments/files/21215016/youtubeUnblock.--trace.txt)
Но понятнее ситуация не стала)

**Reply by bursoft-git** [reply]
Нашёл источник проблемы. У меня на роутере есть так называемые "политики доступа", где можно настроить приоритет подключений, резервирование и всякое такое. Для моих устройств была выбрана такая политика доступа, где сразу использовалось подключение wireguard, а если оно выключено, то подключение ethernet. Стоило мне переключиться на "политику по умолчанию", где первым идёт подключение ethernet, то сразу пошёл трафик в хату) Спасибо за помощь, теперь буду углубляться в параметры)

**Comment by tirealcom** [comment]
Здравствуйте
Роутер ASUS RT-AX59U, подключен к основному как точка доступа
Поставил вашу программу и она работает, но перестаёт работать локальная сеть (rdp, smb, plex)
Что нужно сделать чтобы работало и unblock и локальная сеть?

**Comment by gelioson** [comment]
Доброго вам здравия и огромное спасибо за ваш труд.
Попробовал себе сделать шлюз из малинки по как описано в [каменте](https://github.com/Waujito/youtubeUnblock/discussions/172#discussioncomment-11691976) - работает просто прекрасно. Но отдельное устройство не есть гуд, посему есть великое желание установить youtubeUnblock на роутер.
Проблема в том, что роутер у меня Asus AC68U древней ревизии, который на ядре 2.6. Скачал я у вас вот такой пакетик [youtubeUnblock-1.1.0-1-473af29-entware-armv7-2.6.ipk](https://github.com/Waujito/youtubeUnblock/releases/download/v1.1.0/youtubeUnblock-1.1.0-1-473af29-entware-armv7-2.6.ipk) для armv7-2.6 и он у меня прекрасно установился и даже запустился. Но ютуб, к сожалению, не заработал. На самом роутере тест на гуглвидео проходит, на других компах в той же сети НЕ проходит. По трейсу видно, что пакеты он вроде бы ловит и даже как-то обрабатывает, но результата нет. Флаги что на малинке, что тут одинаковые.
Подскажите, пожалуйста, куда дальше копать?
Прикладываю на всякий 
[youtubeUnblock.--trace.txt](https://github.com/user-attachments/files/21927384/youtubeUnblock.--trace.txt)
[iptables.-t.mangle.-nvL.txt](https://github.com/user-attachments/files/21927388/iptables.-t.mangle.-nvL.txt)

**Comment by olegshulyakov** [comment]
## Установка пакета Youtube Unblock на роутер Keenetic

Предварительно у вас должен быть установлен Entware

- [Установка OPKG Entware на встроенную память роутера](https://help.keenetic.com/hc/ru/articles/360021888880-Установка-OPKG-Entware-на-встроенную-память-роутера)
- [Установка системы пакетов репозитория Entware на USB-накопитель](https://help.keenetic.com/hc/ru/articles/360021214160-Установка-системы-пакетов-репозитория-Entware-на-USB-накопитель)

> [!IMPORTANT]
> Шаги описаны по памяти

### Шаги:
1. Подключиться к роутеру `telnet 192.168.1.1`.
2. Открыть терминал роутера `exec sh`.
3. Обновить список пакетов `opkg update`.
4. Создать папку для скачивания пакета и хранения конфигурации `mkdir -p /storage/home/admin`.
5. Перейти в эту папку `cd /storage/home/admin`.
6. Скачать пакет Youtube Unblock `wget https://github.com/Waujito/youtubeUnblock/releases/download/...` (ссылку взять со страницы [релизов](https://github.com/Waujito/youtubeUnblock/releases))
7. Установить пакет `opkg install youtubeUnblock-1.1.0-1-473af29-entware-mipsel-3.4.ipk`.
8. Проверить статус `/opt/etc/init.d/S91youtubeUnblock status`
9. Создайте файл нужных вам хостов `touch hosts.txt`
10. Отредактируйте его редактором Vi `vi hosts.txt`
11. Пропишите его в параметрах запуска `vi /opt/etc/init.d/S91youtubeUnblock`:
```
# Pass your args here
ARGS="--sni-domains-file=/storage/home/admin/hosts.txt"
```
12. Перезапустите `/opt/etc/init.d/S91youtubeUnblock restart`

### Пример `hosts.txt`:
```
yt.be
youtu.be
youtube.com
googlevideo.com
ggpht.com
ytimg.com
googleapis.com
gstatic.com
googleusercontent.com
l.google.com
play.google.com
```

**Reply by diks13** [reply]
Здравствуйте, сделал все по вашей инструкции, столкнулся с проблемой.
1. Папка admin  создается во внутренней памяти а не на флешке.
2. youtubeUnblock-1.1.0-1-473af29-entware-mipsel-3.4.ipk закинул через файловый менеджер, выполняю команду на установку, пишет установка прошла успешно, но файл S91youtubeUnblock по своему адресу  не появляется. 
Сегодня попробую переустановить Entware.

**Reply by olegshulyakov** [reply]
> Папка admin создается во внутренней памяти а не на флешке.

Поправьте путь `/storage/home/admin` на требуемый вам, он приведен как пример.

> файл S91youtubeUnblock по своему адресу не появляется

Проверьте его в `ls -l /opt/bin/youtubeUnblock`. Если там нет, то скорее всего были какие-то ошибки в выводе установки скрипта.

P.S.
- `opkg list-installed` покажет установленные пакеты.
- Удаление можно выполнить через `opkg remove youtubeUnblockEntware`

**Reply by diks13** [reply]
Здравствуйте, я очень рад вашей поддержке, у меня появилась надежда на запуск.
Когда я начал переустановку я получил ошибку несовместимости, и попробовал другую архитектуру по которой установка прошла успешно.
Giga (KN-1012) EAEU

```
`KeeneticOS version 4.03.C.6.3-9, copyright (c) 2010-2025 Keenetic Ltd.


This software is a subject of Keenetic Ltd. end-user licence agreement. By using it you agree on terms and conditions
hereof. For more information please check https://keenetic.com/legal

(config)> exec sh


BusyBox v1.37.0 (2025-01-13 14:28:09 UTC) built-in shell (ash)

/ # opkg update
Downloading http://bin.entware.net/aarch64-k3.10/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/entware
Downloading http://bin.entware.net/aarch64-k3.10/keenetic/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/keendev
/ # mkdir -p /storage/home/admin
/ # cd /storage/home/admin
/storage/home/admin # opkg install youtubeUnblock-1.1.0-1-473af29-entware-aarch64-3.10.ipk
Installing youtubeUnblockEntware (1.1.0-1-473af29~473af29-r1) to root...
Configuring youtubeUnblockEntware.
/storage/home/admin # /opt/etc/init.d/S91youtubeUnblock status
stopped
/storage/home/admin # /opt/etc/init.d/S91youtubeUnblock restart
 Shutting down youtubeUnblock 
youtubeUnblock 1.1.0-1
Bypasses deep packet inspection systems that rely on SNI

Running with flags: --queue-num=537 --threads=1 --packet-mark=32768 --daemonize --tls=enabled --frag=tcp --frag-sni-reverse=1 --frag-sni-faked=0 --frag-middle-sni=1 --frag-sni-pos=1 --fk-winsize=0 --fake-sni=1 --fake-sni-seq-len=1 --fake-sni-type=default --faking-strategy=pastseq --fake-seq-offset=10000 --seg2delay=0 --sni-domains=<trie of 109 vertexes> --sni-detection=parse --synfake=0 --udp-filter-quic=disabled 
/opt/etc/init.d/S91youtubeUnblock: line 83: iptables: not found
/opt/etc/init.d/S91youtubeUnblock: line 83: iptables: not found
/opt/etc/init.d/S91youtubeUnblock: line 83: iptables: not found
/opt/etc/init.d/S91youtubeUnblock: line 83: iptables: not found
/opt/etc/init.d/S91youtubeUnblock: line 83: ip6tables: not found
/opt/etc/init.d/S91youtubeUnblock: line 83: ip6tables: not found
/opt/etc/init.d/S91youtubeUnblock: line 83: ip6tables: not found
/opt/etc/init.d/S91youtubeUnblock: line 83: ip6tables: not found
 Started youtubeUnblock 
/storage/home/admin # opkg list-installed
busybox - 1.37.0-4
dropbear - 2024.86-1
entware-release - 2025.01-1
findutils - 4.10.0-1
grep - 3.11-1
ldconfig - 2.27-11
libc - 2.27-11
libgcc - 8.4.0-11
libpcre2 - 10.42-1
libpthread - 2.27-11
librt - 2.27-11
libssp - 8.4.0-11
libstdcpp - 8.4.0-11
locales - 2.27-9
opkg - 2024.10.16~38eccbb1-1
opt-ndmsv2 - 1.0-17
poorbox - 1.37.0-1
terminfo - 6.4-3
youtubeUnblockEntware - 1.1.0-1-473af29~473af29-r1
zoneinfo-asia - 2024b-1
zoneinfo-core - 2024b-1
zoneinfo-europe - 2024b-1
/storage/home/admin # `
```

Я получаю ошибку на 83 линии.

```
{
	            ARGV="$@"
	            CMD=$1		# iptables or ip6tables
	             ACTION=$2	# -I, -A, -D
	             shift; shift;
	             RULE="$@"

	             $CMD -C $RULE 2>/dev/null
	             exists=$(( ! $? ))

                     	if [ "$ACTION" = "-A" -o "$ACTION" = "-I" ]
                       then
		                 if [ $exists -eq 0 ]; then
83 Line      			$ARGV || return 1
 		            fi
 	                  else # -D
 		        if [ $exists -ne 0 ]; then
 			$ARGV
 		fi
 	fi
}
```


Ютуб все еще не работает, как это можно исправить ?

**Comment by iunewind** [comment]
У меня Asus (AX86U) Merlin т.е. установлен Entware. Долгое время все работало, причем была особенность, что ipad и android устройства работают идеально а вот на пк очень часто подвисает на какоето время или запускается только после перезагрузки итп. Но пару дней назад все накрылось, ipad перестал работать вообще, хотя андройд и пк работали. Сегодня похоже не работает ничего. 
В обсуждениях вижу, что проблема у многих и ее ктото решает доп настройками в виде параметров запуска и указывает какието скрины с настройками.  

Куда вбивать эти параметры, где эти окна с настройками? В редми про энтвейр полторы строчки. Я правильно понял - настройки доступны только для кинетиков и роутеров с опенврт, асусы идут лесом?

**Reply by iunewind** [reply]
> У меня Asus (AX86U) Merlin т.е. установлен Entware. 

Сам себе отвечу, может кому пригодится - решил задачу через zapret4rocket, это люди сделали скрипт инсталлирующий zapret на "любой" роутер в одну команду (если, разумеется, уже есть openwrt или entware). Для асусов тоже работает, во всяком случае на моем сработало. Ютуб пока починился.

**Reply by ghost** [reply]
Так себе решение. Больше похоже на рекламу другого продукта.

**Reply by iunewind** [reply]
> Так себе решение. Больше похоже на рекламу другого продукта.

Скрипт бесплатный, "рекламирую" тоже бесплатно)) Даже ссылку не стал делать чтобы не приняли за бота если комуто придет в голову такая странная мысль. Однакож комуто пришла)). 

В чем "так себе решение"? Решает туже задачу, что и ютубАнблок, примерна таже сложность установки, только на данный момент у меня youtubeUnblock не работает, и новой версии пока не было, я полистал обсуждения, там люди рандомно решают дописыванием параметров, но варианта дописать параметры для моего роутера или нет, или он для значительно лучше разбирающихся в линухах/коде ребят,  как (и можно ли) это сделать человеку с улицы я не нашел. Скрипт выше заработал (после небольшой пляски) и пока он у меня успешно выполняет задачу. Бонусом - кое какие другие сервисы тоже открыл.
